efi/preinstall/check_tcglog.go

```diff
@@ -0,0 +1,610 @@
+func (r *pcrBankResults) pcrErrs() (out map[tpm2.Handle]error) {
+	out = make(map[tpm2.Handle]error)
+	for pcr, result := range r.pcrs {
+		err := result.Err()
+		if err == nil {
+			continue
+		}
+		out[tpm2.Handle(pcr)] = err
+	}
+	return out
+}
+
+// checkFirmwareLogAgainstTPMForAlg checks the supplied TCG log consistency, reconstructed against the
+// TPM PCRs for the specified algorithm. This only checks TCG defined PCRs (0-7). The supplied
+// mandatoryPcrs argument is required by the returned results so that it can use this to make a decision
+// on whether the bank is ok.
+func checkFirmwareLogAgainstTPMForAlg(tpm *tpm2.TPMContext, log *tcglog.Log, alg tpm2.HashAlgorithmId, mandatoryPcrs tpm2.HandleList) (results *pcrBankResults, err error) {
+	// Check that the TCG log contains the specified algorithm
+	supported := false
+	for _, logAlg := range log.Algorithms {
+		if logAlg != alg {
+			continue
+		}
+		// It does, so that's a good start.
+		supported = true
+		break
+	}
+	if !supported {
+		// The log doesn't contain the specified algorithm
+		return nil, errors.New("digest algorithm not present in log")
+	}
+
+	// Create the result tracker for PCRs 0-7
+	results = newPcrBankResults(alg, mandatoryPcrs)
+
+	seenStartupLocalityEvent := false
+
+	// Iterate over the log
+	for i, ev := range log.Events {
+		if !internal_efi.IsTCGDefinedPCR(ev.PCRIndex) {
+			// Skip all events that aren't defined by the TCG
+			continue
+		}
+
+		if ev.EventType == tcglog.EventTypeNoAction {
+			// EV_NO_ACTION events are informational and not measured, so ignore most of them with the exception
+			// of StartupLocality events, which affects the initial value of PCR0.
+			startupLocalityData, isStartupLocality := ev.Data.(*tcglog.StartupLocalityEventData)
+			if !isStartupLocality {
+				// Not a StartupLocality event.
+				continue
+			}
+
+			// This is the StartupLocality event which is added to the log to indicate
+			// which locality the TPM2_Startup command was executed from. The reference
+			// TPM implementation only allows this to be executed from localities 0 and 3.
+			// The platform may restrict access to locality 3 from the host OS (even from
+			// code running in the kernel). The startup locality affects the reset value of
+			// PCR0 - the least significant byte is set to the startup locality value. PCR0
+			// can also be initialized before TPM2_Startup by a H-CRTM event sequence from
+			// locality 4, which will be reflected in this event in this case by setting the
+			// startup locality to 4, regardless of what locality TPM2_Startup is subsequently
+			// called from.
+			if ev.PCRIndex != internal_efi.PlatformFirmwarePCR {
+				// This event should only ever appear in PCR0
+				results.Lookup(ev.PCRIndex).setErr(errors.New("unexpected StartupLocality event (should be in PCR0)"))
+				continue
+			}
+			if !bytes.Equal(results.Lookup(0).logValue, results.Lookup(0).initialValue) {
```

Comment:
shouldn't this check be a predicate on the struct?

Reply:
Yes, they could be